video/out/vo_gpu_next.c

```diff
@@ -1509,112 +1512,69 @@ static void wait_events(struct vo *vo, int64_t until_time_ns)
-    bool same_cache = false;
-    if (icc_cache && shader_cache)
-        same_cache = strcmp(icc_cache, shader_cache) == 0;
-    if (shader_cache) {
-        FILE *cache = fopen(shader_cache, "rb");
-        if (cache) {
-            int ret = pl_cache_load_file(p->shader_cache, cache);
-            if (same_cache)
-                pl_cache_load_file(p->icc_cache, cache);
-            fclose(cache);
-            if (ret < 0)
-                MP_WARN(p, "Failed loading cache from %s\n", shader_cache);
-        }
-        talloc_free(shader_cache);
-    }
-    if (icc_cache && !same_cache) {
-        FILE *cache = fopen(icc_cache, "rb");
-        if (cache) {
-            int ret = pl_cache_load_file(p->icc_cache, cache);
-            fclose(cache);
-            if (ret < 0)
-                MP_WARN(p, "Failed loading cache from %s\n", icc_cache);
-        }
+    cache->path = mp_path_join(NULL, dir, name);
+    cache->cache = pl_cache_create(pl_cache_params(
+        .log = p->pllog,
+        .max_total_size = max_size,
+    ));
+
+    FILE *file = fopen(cache->path, "rb");
+    if (file) {
+        int ret = pl_cache_load_file(cache->cache, file);
+        fclose(file);
+        if (ret < 0)
+            MP_WARN(p, "Failed loading cache from %s\n", cache->path);
     }
-    talloc_free(icc_cache);
 }
 
-static void save_cache_files(struct priv *p)
+static void cache_uninit(struct priv *p, struct cache *cache)
 {
-    void *ta_ctx = talloc_new(NULL);
-    char *icc_cache = get_cache_file(p, "icc");
-    char *shader_cache = get_cache_file(p, "shader");
-    talloc_steal(ta_ctx, icc_cache);
-    talloc_steal(ta_ctx, shader_cache);
-
-    bool same_cache = false;
-    if (icc_cache && shader_cache)
-        same_cache = strcmp(icc_cache, shader_cache) == 0;
-    for (int i = 0; i < 2; i++) {
-        const char *target_file = i == 0 ? shader_cache : icc_cache;
-        pl_cache target_cache = i == 0 ? p->shader_cache : p->icc_cache;
-
-        if (!target_file)
-            continue;
-        char *tmp = talloc_asprintf(ta_ctx, "%sXXXXXX", target_file);
-        int fd = mkstemp(tmp);
-        if (fd < 0)
-            continue;
-        FILE *cache = fdopen(fd, "wb");
-        if (!cache) {
-            close(fd);
-            continue;
-        }
-        int ret = pl_cache_save_file(target_cache, cache);
-        if (same_cache)
-            ret += pl_cache_save_file(p->icc_cache, cache);
-        fclose(cache);
-        if (ret >= 0)
-            ret = rename(tmp, target_file);
-        if (ret < 0) {
-            MP_WARN(p, "Failed saving cache to %s\n", target_file);
-            unlink(tmp);
-        }
+    char *tmp = NULL;
+    if (!cache->cache)
+        goto done;
 
-        if (same_cache)
-            break;
+    assert(cache->path);
+    tmp = talloc_asprintf(NULL, "%sXXXXXX", cache->path);
+    int fd = mkstemp(tmp);
+    if (fd < 0)
+        goto done;
+    FILE *file = fdopen(fd, "wb");
+    if (!file) {
+        close(fd);
```

Comment:
I know it is unrelated to your changes, but could you add `unlink(tmp);` here? To not leave temp file if `fdopen` fails for whatever reason. I don't think there is a possibility for it to fail in this case, but anyway.

I noticed this the other day, but deem too nonsignificant to send pr. But as you bikeshedding it anyway.

Reply:
Added